### PR TITLE
Forbid unsafe casting in bitwise operation

### PR DESCRIPTION
**Pull request by sklam**
This patch adds a `no_unsafe_casting` flag the `FunctionTemplate` and resolution methods to forbid the selection of signatures that require unsafe casting (e.g. float -> int).  

Bitwise function templates are updated to use this flag.

### CODE REVIEW

**Comment by sklam** [comment]
This should fix #2365

File: numba/tests/test_operators.py [hunk 1]
```diff
@@ -1022,6 +1022,32 @@ def test_bitwise_not(self, flags=force_pyobj_flags):
     def test_bitwise_not_npm(self):
         self.test_bitwise_not(flags=Noflags)
 
+    def test_bitwise_float(self):
+        """
+        Make sure that bitwise float operations are not allowed
+        """
+        def assert_reject_compile(pyfunc, argtypes):
+            msg = 'expecting TypingError with compiling {}'.format(pyfunc)
```
Comment (stuartarchibald) [inline numba/tests/test_operators.py]: `s/with/when/`

File: numba/tests/test_operators.py [hunk 2]
```diff
@@ -1022,6 +1022,32 @@ def test_bitwise_not(self, flags=force_pyobj_flags):
...
+            with self.assertRaises(errors.TypingError, msg=msg):
+                compile_isolated(pyfunc, argtypes)
```
Comment (stuartarchibald) [inline numba/tests/test_operators.py]: Could the exception message be checked to assert it contains `invalid usage of <op>`?

File: numba/typing/builtins.py [hunk 3]
```diff
@@ -246,10 +246,14 @@ class BitwiseShiftOperation(ConcreteTemplate):
...
-             for op in sorted(types.signed_domain)]
-    cases += [signature(max(op, types.uintp), op, types.intp)
-              for op in sorted(types.unsigned_domain)]
+    cases = [signature(max(op, types.intp), op, op2)
+             for op in sorted(types.signed_domain)
+             for op2 in [types.uintp, types.intp]]
+    cases += [signature(max(op, types.uintp), op, op2)
+              for op in sorted(types.unsigned_domain)
+              for op2 in [types.uintp, types.intp]]
+    no_unsafe_casting = True
```
Comment (stuartarchibald) [inline numba/typing/builtins.py]: Should this be `unsafe_casting = False` and then defaults to `True` in the overload resolution function defaults, can't decide which spelling sounds right!?

Comment (sklam) [inline numba/typing/builtins.py]: yeah, probably without the `no_` is better.  One less negation when thinking about it.

**Comment by stuartarchibald** [comment]
Thanks for the fixes. This looks good. Merging.

**Pull request merged**